### ISSUE TITLE
`tofu test` - allow provider blocks in tests to access variables and run outputs

### ISSUE THREAD

**Issue by RLRabinowitz** [opentofu/opentofu#1185]
### OpenTofu Version

```shell
1.6.0
```

### Use Cases

Like in the main tofu configuration, there are cases where we'd like to initialize providers based on dynamic configuration in the tests. 

One example is referring to variables that have been defined in the test file. That would allow to test the module with different regions for a provider, for example
Another example is referring to run outputs - Running some setup configuration, which might get some secret AWS credentials, and then you could initialize the provider with these credentials

### Attempted Solutions

Currently that's not possible, and the provider configuration in test suites is pretty static

### Proposal

Allow `provider` blocks in test files to refer to variables defined in the test file, and also allow them to refer to `run` block outputs:

```
variables {
  aws_access_key_id = "aa"
  aws_secret_access_key = "bb"
}

provider "aws" {
  access_key = var.aws_access_key_id
  secret_key = var.aws_secret_access_key
}
```

and:
```
provider "aws" {
  access_key = run.setup.aws_access_key_id
  secret_key = run.setup.aws_secret_access_key
}

run "setup" {
  module {
    source = "./testing/setup"
  }
}
```

### References

Related issue in the Terraform Repository - https://github.com/hashicorp/terraform/issues/34007

**Comment by mmigliari** [comment]
I would love to see this. It can make deploying in CI/CD on certain situations a lot more straightforward.

**Comment by sanskruti-shahu** [comment]
Hey, I would like to take up this issue.

**Comment by sanskruti-shahu** [comment]
Hey @Yantrio, I would like to work upon this issue.

**Comment by sanskruti-shahu** [comment]
Hey, I would like to take up this issue.

**Comment by simonostendorf** [comment]
> [@RLRabinowitz](https://github.com/RLRabinowitz) I've just done some preliminary tests, is it possible that this is already fixed?

I dont think so, I still get the error `A managed resource "run" "<run-block-name>" has not been declared in the root module.` when I try to use run block outputs to init providers.

**Comment by LarsSven** [comment]
Is this still being worked on? We are really missing this feature and would love it to be supported.

**Comment by simonostendorf** [comment]
> Is this still being worked on? We are really missing this feature and would love it to be supported.

I really would like to see this too, this would be a game changer for me.
